pydantic/_internal/_fields.py

```diff
@@ -158,7 +159,18 @@ def collect_model_fields(  # noqa: C901
         if _typing_extra.is_classvar_annotation(ann_type):
             class_vars.add(ann_name)
             continue
-        if _is_finalvar_with_default_val(ann_type, getattr(cls, ann_name, PydanticUndefined)):
+
+        assigned_value = getattr(cls, ann_name, PydanticUndefined)
+
+        if _is_finalvar_with_default_val(ann_type, assigned_value):
+            warnings.warn(
+                f'Annotation {ann_name!r} is marked as final and has a default value. Pydantic treats {ann_name!r} as a '
+                f'class variable, but it will be considered as an normal field in V3. If you still want {ann_name!r} to be '
+                'considered as a class variable, annotate it as: `ClassVar[<type>] = <default>.`',
```

Comment:
Can we link to the typing spec change that motivated this change?

Reply:
I added "to be aligned with dataclasses" as the typing spec change was actually just an update to follow what stdlib dataclasses where doing.